src/main/java/htsjdk/samtools/cram/build/SliceFactory.java

```diff
@@ -0,0 +1,333 @@
+                    throw new CRAMException("Invalid coord-sorted input - unmapped records must be last");
+                } else {
+                    // Not coordinate sorted, and we're going from unmapped to mapped, so allow the mapped
+                    // record into the same slice with the unmapped ones, since there is no index query
+                    // concern since we're not coord sorted anyway (though there is no reference compression
+                    // happening in this container).
+                    return numberOfSAMRecords >= maxRecordsPerSlice ?
+                            ReferenceContext.UNINITIALIZED_REFERENCE_ID :
+                            ReferenceContext.MULTIPLE_REFERENCE_ID;
+                }
+
+            case ReferenceContext.MULTIPLE_REFERENCE_ID:
+                // This always accumulates a full multi-ref slice, but maybe it would be more efficient
+                // to emit smaller multi-ref slices on the theory that the stream will get back on track
+                // for single ref, at least for coord-sorted.
+                if (coordinateSorted) {
+                    return numberOfSAMRecords < minimumSingleReferenceSliceThreshold ?
+                            ReferenceContext.MULTIPLE_REFERENCE_ID :
+                            ReferenceContext.UNINITIALIZED_REFERENCE_ID; // emit a small mutli-ref
+                } else {
+                    // multi-ref, not coord sorted
+                    return numberOfSAMRecords >= maxRecordsPerSlice ?
+                            ReferenceContext.UNINITIALIZED_REFERENCE_ID :
+                            ReferenceContext.MULTIPLE_REFERENCE_ID;
+                }
+
+            default:
+                // So far everything we've accumulated for the next slice is on a single reference contig
+                if (nextReferenceIndex == currentReferenceContext) {
+                    // still on the same reference contig
+                    return numberOfSAMRecords >= maxRecordsPerSlice ?
+                            ReferenceContext.UNINITIALIZED_REFERENCE_ID :
+                            nextReferenceIndex;
+                } else {
+                    // switching to either a new reference contig, or to unmapped
+                    return numberOfSAMRecords < minimumSingleReferenceSliceThreshold ?
+                            // if we already have accumulated at least one slice, then we emit it rather than
+                            // switch to multi-ref so we can prevent a multi-ref slice from being packed into
+                            // a container with a single-ref slice (which violates the spec, so to do so
+                            // would require making both slices multi-ref)
+                            getNumberOfSliceEntries() > 0 ?
+                                    ReferenceContext.UNINITIALIZED_REFERENCE_ID:
+                                    ReferenceContext.MULTIPLE_REFERENCE_ID :
```

Comment:
Parenthesize the inner `? :` statement for clarity.

Reply:
Done.